Posts/2022/08/many-wmi-flavours.md

```diff
@@ -0,0 +1,390 @@
+In the other hand, querying and doing more complex operations can be time-consuming, and can involve more objects to keep track of.
+  
+- Using the _WMI Scripting API_ is great when you have to build whole scripts to manage WMI. Once you have the _SWbemServices_ object you can work with pretty much anything else.  
+It is also rewarding performance-wise, compared to the previous method, since you are working with the RCW interfaces directly. The _System.Management_ namespace will wrap these interfaces to provide abstraction.  
+But this comes at a cost. If you want to retrieve single objects or perform queries to analyze data, this method can be a little annoying to work with.  
+  
+- The _CIM Cmdlets_ are number one on performance when querying multiple-object datasets. The _CimInstance_ object is great to work with, specially when combining with other known objects, like the _PSCustomObject_.  
+In the other hand, calling methods is not as straight forward as on the previous methods. And to access methods like _Put()_, _Get()_ or _Delete()_ can be challenging.
+
+## Bonus
+
+Let's create our own Namespace under root, and implement a custom class!
+We will use the _System.Management_ namespace, but now you can use what you learn to implement this using the other methods as well.
+
+```powershell-interactive
+PS>_ $namespace = ([wmiclass]'root:__Namespace').CreateInstance()
+PS>_ $namespace.Name = 'ScriptingBlogCoolNamespace'
+PS>_ $namespace.Put()
+
+Path          : \\.\root:__NAMESPACE.Name="ScriptingBlogCoolNamespace"
+RelativePath  : __NAMESPACE.Name="ScriptingBlogCoolNamespace"
+Server        : .
+NamespacePath : root
+ClassName     : __NAMESPACE
+IsClass       : False
+IsInstance    : True
+IsSingleton   : False
+
+PS>_ $class = [wmiclass]::new('root\ScriptingBlogCoolNamespace', '', $null)
+PS>_ $class['__Class'] = 'CustomClass'
+PS>_ $class.Qualifiers.Add('Static', $true)
+PS>_ $class.Properties.Add('Source', 'Custom class with .NET!')
+PS>_ $class.Properties.Add('PropertyKey', 1)
+PS>_ $class.Properties['PropertyKey'].Qualifiers.Add('Key', $true) ## You need a Key property, otherwise WMI wouldn't be able to assemble the path of a new instance.
+PS>_ $class.Put()
+
+Path          : \\.\root\ScriptingBlogCoolNamespace:CustomClass
+RelativePath  : CustomClass
+Server        : .
+NamespacePath : root\ScriptingBlogCoolNamespace
+ClassName     : CustomClass
+IsClass       : True
+IsInstance    : False
+IsSingleton   : False
+
+PS>_ $instance = ([wmiclass]'root\ScriptingBlogCoolNamespace:CustomClass').CreateInstance()
+PS>_ $instance.Source = 'CustomInstance!'
+PS>_ $instance.Put()
+
+Path          : \\.\root\ScriptingBlogCoolNamespace:CustomClass.PropertyKey=1
+RelativePath  : CustomClass.PropertyKey=1
+Server        : .
+NamespacePath : root\ScriptingBlogCoolNamespace
+ClassName     : CustomClass
+IsClass       : False
+IsInstance    : True
+IsSingleton   : False
+```
```

Comment:
````suggestion
```powershell
$namespace = ([wmiclass]'root:__Namespace').CreateInstance()
$namespace.Name = 'ScriptingBlogCoolNamespace'
$namespace.Put()
```

```Output
Path          : \\.\root:__NAMESPACE.Name="ScriptingBlogCoolNamespace"
RelativePath  : __NAMESPACE.Name="ScriptingBlogCoolNamespace"
Server        : .
NamespacePath : root
ClassName     : __NAMESPACE
IsClass       : False
IsInstance    : True
IsSingleton   : False
```

```powershell
$class = [wmiclass]::new('root\ScriptingBlogCoolNamespace', '', $null)
$class['__Class'] = 'CustomClass'
$class.Qualifiers.Add('Static', $true)
$class.Properties.Add('Source', 'Custom class with .NET!')
$class.Properties.Add('PropertyKey', 1)
## You need a Key property, otherwise WMI wouldn't be able to assemble the path of a new instance.
$class.Properties['PropertyKey'].Qualifiers.Add('Key', $true)
$class.Put()
```

```Output
Path          : \\.\root\ScriptingBlogCoolNamespace:CustomClass
RelativePath  : CustomClass
Server        : .
NamespacePath : root\ScriptingBlogCoolNamespace
ClassName     : CustomClass
IsClass       : True
IsInstance    : False
IsSingleton   : False
```

```powershell
$instance = ([wmiclass]'root\ScriptingBlogCoolNamespace:CustomClass').CreateInstance()
$instance.Source = 'CustomInstance!'
$instance.Put()
```

```Output
Path          : \\.\root\ScriptingBlogCoolNamespace:CustomClass.PropertyKey=1
RelativePath  : CustomClass.PropertyKey=1
Server        : .
NamespacePath : root\ScriptingBlogCoolNamespace
ClassName     : CustomClass
IsClass       : False
IsInstance    : True
IsSingleton   : False
```
````